src/hdmf/data_utils.py

```diff
@@ -20,7 +20,15 @@ def append_data(data, arg):
         data.append(arg)
         return data
     elif isinstance(data, np.ndarray):
-        return np.append(data,  np.expand_dims(arg, axis=0), axis=0)
+        if isinstance(arg, np.ndarray):
+            if data.ndim == arg.ndim:
```

Comment:
Note that with this change, when `data = np.array([[1, 2, 3], [4, 5, 6]])` and `arg = np.array([[7, 8, 9]])`, this code used to raise an error. Now it flattens the list to `array([1, 2, 3, 4, 5, 6, 7, 8, 9])`. 

`np.append` works for structured arrays here because we happen to be dealing only with 1D structured arrays, so flattening the array is OK - it is still 1D.

It looks like with these changes, you are trying to capture the case where you call `append_data` on:
```python
c_data = np.array([('Homo sapiens', 24)], dtype=[('species', 'U50'), ('age', 'i4')])
c_data2 = np.array([('Mus musculus', 24)], dtype=[('species', 'U50'), ('age', 'i4')])
```

But if we want to be consistent with past behavior, that _should_ raise an error, because `append_data` is designed to add one new row, and `extend_data` is designed to add multiple rows (this mimics python list append/extend)

To test adding one new compound dtype row, the test should be
```python
c_data = np.array([('Homo sapiens', 24)], dtype=[('species', 'U50'), ('age', 'i4')])
c_data2 = np.array(('Mus musculus', 24), dtype=[('species', 'U50'), ('age', 'i4')])
```
The unmodified `append_data` handles this case just fine. 

With these changes, we are adding new behavior that is inconsistent between structured arrays and non-structured arrays, and can result in silent unlikely-to-be-intended flattening for non-structured arrays. That all seems unnecessary.

I suggest, either
1. revert all changes to this function and change the test
2. use a more specific test for structured arrays like len(dtype) to add new behavior that allows a user to append a structured array of multiple elements to a structured array of multiple elements.

Reply:
I'm removing my idea of supporting appending arrays.